pkg/capabilities/triggers/mercury_trigger.go

```diff
@@ -0,0 +1,256 @@
+	if err != nil {
+		return err
+	}
+
+	ch, ok := o.chans[wid+triggerID]
+	if ok {
+		close(ch)
+	}
+
+	for _, feedID := range o.feedIDsForTriggerID[triggerID] {
+		o.triggerIDsForFeedID[feedID] = removeFromSlice(o.triggerIDsForFeedID[feedID], triggerID)
+	}
+
+	delete(o.chans, wid+triggerID)
+	delete(o.feedIDsForTriggerID, triggerID)
+	delete(o.triggerIDToWorkflowID, triggerID)
+
+	return nil
+}
+
+// Get array of feedIds from CapabilityRequest req
+func (o *MercuryTriggerService) GetFeedIDs(req capabilities.CapabilityRequest) []int64 {
+	feedIDs := make([]int64, 0)
+	// Unwrap the inputs which should return pair (map, nil) and then get the feedIds from the map
+	if inputs, err := req.Inputs.Unwrap(); err == nil {
+		if feeds, ok := inputs.(map[string]interface{})["feedIds"].([]any); ok {
+			// Copy to feedIds
+			for _, feed := range feeds {
+				if id, ok := feed.(int64); ok {
+					feedIDs = append(feedIDs, id)
+				}
+			}
+		}
+	}
+	return feedIDs
+}
+
+// Get the triggerId from the CapabilityRequest req map
+func (o *MercuryTriggerService) GetTriggerID(req capabilities.CapabilityRequest) (string, error) {
+	// Unwrap the inputs which should return pair (map, nil) and then get the triggerId from the map
+	inputs, err := req.Inputs.Unwrap()
+	if err != nil {
+		return "", err
+	}
+	if id, ok := inputs.(map[string]interface{})["triggerId"].(string); ok {
+		return id, nil
+	}
+
+	return "", fmt.Errorf("triggerId not found in inputs")
+}
+
+func GenerateTriggerEventID(reports []mercury.FeedReport) string {
+	// Let's hash all the feedIds and timestamps together
+	sort.Slice(reports, func(i, j int) bool {
+		if reports[i].FeedID == reports[j].FeedID {
+			return reports[i].ObservationTimestamp < reports[j].ObservationTimestamp
+		}
+		return reports[i].FeedID < reports[j].FeedID
+	})
+	s := ""
+	for _, report := range reports {
+		s += strconv.FormatInt(report.FeedID, 10) + strconv.FormatInt(report.ObservationTimestamp, 10) + ","
+	}
+	return sha256Hash(s)
+}
+
+func ValidateInput(mercuryTriggerEvent values.Value) error {
+	// TODO: Fill this in
+	return nil
+}
+
+func ExampleOutput() (values.Value, error) {
+	event := mercury.TriggerEvent{
+		TriggerType: "mercury",
+		ID:          "123",
+		Timestamp:   "2024-01-17T04:00:10Z",
+		Payload: []mercury.FeedReport{
+			{
+				FeedID:               2,
+				FullReport:           []byte("hello"),
+				BenchmarkPrice:       100,
+				ObservationTimestamp: 123,
+			},
+			{
+				FeedID:               3,
+				FullReport:           []byte("world"),
+				BenchmarkPrice:       100,
+				ObservationTimestamp: 123,
+			},
+		},
+	}
+	return mercury.Codec{}.WrapMercuryTriggerEvent(event)
+}
+
+func ValidateConfig(config values.Value) error {
+	// TODO: Fill this in
+	return nil
+}
+
+func sha256Hash(s string) string {
+	hash := sha256.New()
+	hash.Write([]byte(s))
+	return hex.EncodeToString(hash.Sum(nil))
+}
+
+func removeFromSlice(list []string, element string) []string {
```

Comment:
Maybe change the slice to a set (map[string]struct{})? Then you wouldn't need to have this helper.